### PR TITLE
Issue #2910630 Add support for fuzziness

### PR DESCRIPTION
**Pull request by m4olivei**
Resolves <a href="http://drupal.org/node/2910630">Issue #2910630</a>. This PR adds support for fuzziness.

It does a couple things:

* Adds a setting to the backend configuration so let an admin set the fuzziness.
* Adds a dependency on my fork of the `makinacorpus/php-lucene` package,
 which is a nice little API for building up Lucene queries (https://github.com/m4olivei/php-lucene-query).
* Incorporates the new fuzziness setting and the Lucene query builder API into the module.

To test:

1. Install and/or composer update the elasticsearch_connector module.
1. Set the fuzziness on your Elasticsearch Search API server (eg. /admin/config/search/search-api/server/<machine_name>/edit`
1. Setup a view and/or Search API pages page, and issue a keyword search. Fuzziness should be added at the specified distance.

### CODE REVIEW

File: src/ElasticSearch/Parameters/Builder/SearchBuilder.php
Comment (deviantintegral) [inline src/ElasticSearch/Parameters/Builder/SearchBuilder.php:256]: Why do we do this?

File: src/Plugin/search_api/backend/SearchApiElasticsearchBackend.php
Comment (deviantintegral) [inline src/Plugin/search_api/backend/SearchApiElasticsearchBackend.php]: What does "auto" mean? Are there upstream docs we can link to?

File: src/Plugin/search_api/backend/SearchApiElasticsearchBackend.php
Comment (deviantintegral) [inline src/Plugin/search_api/backend/SearchApiElasticsearchBackend.php]: Does it make sense to default fuzziness to enabled?

**Review by deviantintegral — COMMENTED** [review]
I tested this locally and it works as promised. Just a few questions in the code.

File: src/ElasticSearch/Parameters/Builder/SearchBuilder.php
Comment (m4olivei) [reply src/ElasticSearch/Parameters/Builder/SearchBuilder.php:256]: Because of how we get `$keys` from Search API module. As per docblock, `$keys` is documented in `\Drupal\search_api\ParseMode\ParseModeInterface::parseInput()`. Keys can be nested, which each level optionally defining a `#conjunction` and/or `#negation`, we want to loop over the keys (string or array) themselves only, after saving the property values above, we don't care about them. It's kind of a gnarly data structure I know. D7 arrays live on!

File: src/Plugin/search_api/backend/SearchApiElasticsearchBackend.php
Comment (m4olivei) [reply src/Plugin/search_api/backend/SearchApiElasticsearchBackend.php]: Why yes, yes there are. Will update.

File: src/Plugin/search_api/backend/SearchApiElasticsearchBackend.php
Comment (m4olivei) [reply src/Plugin/search_api/backend/SearchApiElasticsearchBackend.php]: Now that you mention it, yeah, it kinda does. I think I was being conservative, going for an opt in style feature update, but this module doesn't have a release yet on the 8.x-5.x branch, so lets do it.

**Comment by juampynr** [comment]
@m4olivei, this works but the problem I see with it is that it makes it impossible to disable fuzziness for a particular filter, right?

Can't this be something that we add to each Views filter instead? Is that even possible with the existing APIs?

**Comment by m4olivei** [comment]
@juampynr humm, depends what you mean. Fuzziness only applies to the full text search, not the other filters. As this stands, we apply fuzzy matching to each keyword in the full text search. Technically you could apply fuzziness to individual keywords within the full text search string, eg. if the search was for "quick brown", you could turn the search query_string into "quick~ or brown", which would be fuzzy for the word "quick", but exact matching for "brown". There isn't any good way to make that configurable though.

However, if your talking about turning fuzziness on/off per search page, that might be desirable, and would be doable as you suggest, through a setting on the view, or views filter, and a setting on the search API page. It starts to get more complex though, maintaining settings on the views filter and the search API page. That's why I started here. I figured this covers most use cases, where a site would typically only setup a single search page (views or search API pages), in which they do or don't want fuzzy matching.

File: src/ElasticSearch/Parameters/Builder/SearchBuilder.php
Comment (deviantintegral) [reply src/ElasticSearch/Parameters/Builder/SearchBuilder.php:256]: Ah, got it! Since $keys is defined in ParseModeInterface, should we file an issue to add a "filter" method to that interface and implementation? Either way, certainly doesn't block this PR, it just seems a little confusing to jump around like that.

**Review by deviantintegral — APPROVED** [review]

**Comment by juampynr** [comment]
Yeah, I meant to do it at the Views filter level so it would be a setting listed here:

image

However, I see your point and this already solves most of the scenarios so I am okay with this pull request.

**Comment by nignatov** [comment]
What I'm thinking for this PR is actually to integrate it with des-connector library, instead of directly depends on the module itself.

The idea behind des-connector library is be a kind of abstraction between D8 8.x-5.x and 7.x-5.x. 
The idea was born because it is a big overhead to support both versions, when we want to achieve more or less the same things on both versions.

For example in this case, we can wrap the fuzziness in a des-connector class we can easily use in D7.

@m4olivei @juampynr @deviantintegral what do you think about this?

**Comment by juampynr** [comment]
I am okay with that but I am unsure of how much of this PR can be moved there.

**Comment by m4olivei** [comment]
@nignatov that sounds like a great idea. However from my limited exposure to the DESConnector, it doesn't seem like it's there yet to simply add fuzziness support to it. Seems like you'd need to write a large base of query building stuff to it for adding fuzziness to make sense right? If that's true, maybe we can go with this as is, and then refactor it into DESConnecter when there is a better base of query building code in there? What do you think?

Speaking of which, you've mentioned adding a query builder to DESConnector a couple times (at least that's what I now assume you meant). Could you articulate your thoughts and plans wrt to that in a Drupal.org issue, or point me to one? It would be good to put that down and talk it over, as it sounds like a big-ish change.

**Comment by juampynr** [comment]
Agree with @m4olivei.

@nignatov should new changes be proposed at Drupal.org or here? It may confuse others if we use both.

**Comment by nignatov** [comment]
@m4olivei yes, you are correct. I mentioned building the query builder several times but unfortunately no time so far but this is a talk for a beer :). 
Yes, it will be good idea to open an issue about this and start collaborating. I personally prefer github but drupal.org can also works for me. Honestly I'm not sure which one should be the better one. 
However, I agree we can merge this one and improve this when we have the query builder in the DES-Connector.

**Comment by m4olivei** [comment]
@nignatov I think more Drupal people are likely to see and collaborate if you put the issue on Drupal.org. I maintain the Facebook Instant Articles module, and what works for us is to do the issues on drupal.org, and the coding on Github. Pull requests make it easier and better for review. Issues on d.o leave room for contribution credits (https://www.drupal.org/drupalorg/blog/a-guide-to-issue-credits-and-the-drupal.org-marketplace).

**Comment by nignatov** [comment]
@m4olivei really valid points. I like this approach so lets do it like this.
I will close the issues for this project, just to be more obvious to create them in drupal.org

**Comment by m4olivei** [comment]
👍 

You might also consider a `CONTRIBUTING.md` file, here is what we have for FBIA: https://github.com/BurdaMagazinOrg/module-fb_instant_articles/blob/8.x-2.x/CONTRIBUTING.md

This makes is a little more clear the setup you want contributors to follow. See https://help.github.com/articles/setting-guidelines-for-repository-contributors/

**Pull request merged**
